### PR TITLE
[RFC] Sklearn regression

### PR DESCRIPTION
**Pull request by thomaspinder**
## Type of changes

- [ ] Bug fix
- [X] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [ ] I've added tests for new code.
- [ ] I've added docstrings for the new code.

## Description

**Opening a draft PR to receive feedback on the overall API design. Once this has been agreed upon, tests and more detailed typing will be added to the PR along with functionality for classification and optimisation. To prevent a single monstrous PR, I'll have one PR each for regression, classification, and optimisation that each go into the `sklearn_api` branch which can eventually be merged into `main`.**

This PR introduces an SKLearn API for GPJax that allows users to invoke GP modelling through `.fit` and `.predict` commands, as per the SKLearn API. Further to this, the ability to _score_ the GP model is introduced. In this PR, I would like comments around the high-level design choices and suggestions for how it can be improved.

Issue Number: N/A

### CODE REVIEW

File: docs/examples/sklearn_api.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+# ## Dataset
+#
+# With the necessary modules imported, we simulate a dataset
+# $\mathcal{D} = (\boldsymbol{x}, \boldsymbol{y}) = \{(x_i, y_i)\}_{i=1}^{100}$ with inputs $\boldsymbol{x}$
+# sampled uniformly on $(-3., 3)$ and corresponding independent noisy outputs
+#
+# $$\boldsymbol{y} \sim \mathcal{N} \left(\sin(4\boldsymbol{x}) + \cos(2 \boldsymbol{x}), \textbf{I} * 0.3^2 \right).$$
+#
+# We store our data $\mathcal{D}$ as a GPJax `Dataset` and create test inputs and labels
+# for later.
```
Comment (daniel-dodd) [inline docs/examples/sklearn_api.py:37]: We aren't using the `Dataset` abstraction in this example.

File: docs/examples/sklearn_api.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+# We'll now proceed to build our model. Within the SKLearn API we have three main classes: `GPJaxRegressor`, `GPJaxClassifier`, and `GPJaxOptimizer`/`GPJaxOptimiser`. We'll consider a problem where the response is continuous and so we'll use the `GPJaxRegressor` class. The problem is identical to the one considered in the [Regression notebook](regression.py); however, we'll now use the SKLearn API to build our model. This offers an alternative to the lower-level API and is designed to be similar to the API of [scikit-learn](https://scikit-learn.org/stable/).
+
+# %%
+model = gpx.sklearn.GPJaxRegressor(kernel=gpx.kernels.RBF())
+
+# %% [markdown]
+# Let's now fit the model. Using the abstraction provided by `GPJaxRegressor`, this can be achieved by simply invoking the `fit` method.
+
+# %%
+model.fit(x, y, key=key)
```
Comment (daniel-dodd) [inline docs/examples/sklearn_api.py:71]: Nice. Just to be a _devil's advocate_ here. Do we want to be JAX'y or Sklearn'y with regards to _random_state_. I.e., in sklearn regressor you would provide an integer (seed like) _random_state_ input. Do we want to import a key from jax.random and follow the api of key passing, or do we just want to do `model.fit(x,y)` with an argument `random_state: int = 0` or something?

File: docs/examples/sklearn_api.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+# %% [markdown]
+# ## Model evaluation
+#
+# With a model now fit, we wish to evaluate the model's quality. To do this, we can _score_ the model using either a GPJax scoring function, or a [metric from sklearn](https://scikit-learn.org/stable/modules/model_evaluation.html). We demonstrate how this can be achieved using GPJax's log-posterior density, and the mean-squared error function from sklearn. Scoring functions can be evaluation on either the training or test set, as we show below.
+
+# %%
+from sklearn.metrics import mean_squared_error
+
+model.score(xtest, ytest, gpx.sklearn.SKLearnScore("mse", mean_squared_error))
+model.score(x, y, gpx.sklearn.LogPredictiveDensity())
```
Comment (daniel-dodd) [inline docs/examples/sklearn_api.py:90]: Looks clean!

File: docs/examples/sklearn_api.py (hunk 2 above)
Comment (thomaspinder) [reply docs/examples/sklearn_api.py:71]: Hmm, straight out the gate, I lean towards using a key, just as it guarantees reproducibility inside larger codebases. What are your thought?

File: docs/examples/sklearn_api.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+ax.plot(xtest, mu, label="Predictive mean", color=cols[1])
+ax.legend(loc="center left", bbox_to_anchor=(0.975, 0.5))
+
+
+# %% [markdown]
+# ## System configuration
+
+# %%
+# %reload_ext watermark
+# %watermark -n -u -v -iv -w -a 'Thomas Pinder & Daniel Dodd'
```
Comment (daniel-dodd) [inline docs/examples/sklearn_api.py:140]: Thanks! But this is not my work. 😜
```suggestion
# %reload_ext watermark
# %watermark -n -u -v -iv -w -a 'Thomas Pinder'
```

File: docs/examples/sklearn_api.py (hunk 1 above)
Comment (thomaspinder) [reply docs/examples/sklearn_api.py:37]: Oop, copy-and-paste hangover. Thanks!

File: gpjax/sklearn/strategy.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@dataclass
+class VariationalInference(AbstractStrategy):
+    pass
+
+
+@dataclass
+class MCMCInference(AbstractStrategy):
+    pass
```
Comment (daniel-dodd) [inline gpjax/sklearn/strategy.py:21]: Is this file used?

File: docs/examples/sklearn_api.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+fig, ax = plt.subplots()
+ax.plot(x, y, "o", label="Observations", color=cols[0])
+ax.plot(xtest, ytest, label="Latent function", color=cols[1])
+ax.legend(loc="best")
+
+# %% [markdown]
+# ## Model building
+#
+# We'll now proceed to build our model. Within the SKLearn API we have three main classes: `GPJaxRegressor`, `GPJaxClassifier`, and `GPJaxOptimizer`/`GPJaxOptimiser`. We'll consider a problem where the response is continuous and so we'll use the `GPJaxRegressor` class. The problem is identical to the one considered in the [Regression notebook](regression.py); however, we'll now use the SKLearn API to build our model. This offers an alternative to the lower-level API and is designed to be similar to the API of [scikit-learn](https://scikit-learn.org/stable/).
```
Comment (daniel-dodd) [inline docs/examples/sklearn_api.py:62]: Does `GPJaxClassifier` exist?

File: gpjax/sklearn/optim.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+from gpjax.sklearn.base import BaseEstimator
+
+tfd = tfp.distributions
+
+
+@dataclass
+class GPJaxOptimizer(BaseEstimator):
+    kernel: AbstractKernel
+    mean_function: AbstractMeanFunction = None
+    n_inducing: int = -1
```
Comment (daniel-dodd) [inline gpjax/sklearn/optim.py:16]: What is the intended usage for the `GPJaxOptimizer`? (Why do we have it?)

File: docs/examples/sklearn_api.py (hunk 6 above)
Comment (thomaspinder) [reply docs/examples/sklearn_api.py:62]: Not yet. Will implement once we’re aligned on an API

File: gpjax/sklearn/strategy.py (hunk 5 above)
Comment (thomaspinder) [reply gpjax/sklearn/strategy.py:21]: Not yet. Tbh, accidentally committed this.

File: gpjax/sklearn/optim.py (hunk 7 above)
Comment (thomaspinder) [reply gpjax/sklearn/optim.py:16]: Once we’re happy with an API, I see it being analogous to Scipy’s minimise fn

File: gpjax/sklearn/config.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    full_data: Num[Array, "N D"], n_inducing: int
+) -> Num[Array, "M D"]:
+    return jnp.asarray(kmeans2(full_data, n_inducing, minit="points")[0])
+
+
+@dataclass
+class DefaultConfig:
+    sparse_threshold: Optional[int] = 2000
+    stochastic_threshold: Optional[int] = 20000
+    min_num_inducing: Optional[int] = 100
```
Comment (daniel-dodd) [inline gpjax/sklearn/config.py:41]: Out of interest. How were these numbers chosen? e.g., In the some of the documentation we recommend `5,000` datapoints is still fine for the regression before switching to sparse strategies.

Comment (thomaspinder) [reply gpjax/sklearn/config.py:41]: Adhoc for now. Happy to bump the default to 5000

File: docs/examples/sklearn_api.py (hunk 2 above)
Comment (st--) [reply docs/examples/sklearn_api.py:71]: Fitting a simple GP shouldn't need any randomness though?

I would argue that _if you want to implement the Sklearn API_, you should actually _implement the API_. Which means it should pass only X and y. Otherwise, it wouldn't actually be possible to write some code that takes an arbitrary model and fits it (and does some more stuff with it). If there's any parameters - like a random state - that need to be passed in, these should go in the constructor of the sklearn-style model class.
